Various tweaks to redesigned structures

evanlinjin opened the pull request:
### Description

The following tweaks are made:

* `TxGraph::try_get_chain_position` used to always exclude unconfirmed transactions with last_seen value of 0. However, what is the point of including a transaction in the graph if it cannot be part of the chain history? Additionally, maybe sometimes we don't wish to use the last_seen field at all. The new behavior will consider unconfirmed transactions with last_seen of 0.
* Introduce `LocalChain::insert_block`. This is necessary as a `LocalChain` is updated with another `LocalChain`. The update `LocalChain` needs a simple way to insert blocks.
* `BlockId` should not implement `Anchor`. Discussion: https://github.com/bitcoindevkit/bdk/pull/926#discussion_r1181373238.
* I have removed `TxGraph::relevant_heights` as it is ambiguous. A simpler and more flexible alternative method `TxGraph::all_anchors` is introduced.

### Changelog notice

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
[chain_redesign] Change behavior of `try_get_chain_position`

`TxGraph::try_get_chain_position` used to always exclude unconfirmed
transactions with last_seen value of 0. However, what is the point of
including a transaction in the graph if it cannot be part of the chain
history? Additionally, maybe sometimes we don't wish to use the
last_seen field at all.

The new behavior will consider unconfirmed transactions with last_seen
of 0.

evanlinjin committed:
[chain_redesign] Add `LocalChain::insert_block`

evanlinjin committed:
[chain_redesign] Relax generic constraints

evanlinjin committed:
[chain_redesign] `BlockId` should not implement `Anchor`

If `BlockId` implements `Anchor`, the meaning is ambiguous. We cannot
tell whether it means the tx is anchors at the block, or whether it also
means the tx is confirmed at that block.

Instead, `ConfirmationHeightAnchor` and `ConfirmationTimeAnchor` structs
are introduced as non-ambiguous `Anchor` implementations.

Additionally, `TxGraph::relevant_heights` is removed because it is also
ambiguous. What heights are deemed relevant? A simpler and more flexible
method `TxGraph::all_anchors` is introduced instead.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -85,8 +85,23 @@ impl LocalChain {
+    /// Get the most recent blocks of the best chain.
+    ///
+    /// `max_count` is the upper bound of the number of elements to return.
+    pub fn recent_blocks_of_best_chain(&self, max_count: usize) -> BTreeMap<u32, BlockHash> {
```
Seems like another premature convenience method? It's not used in this PR so maybe wait until the PR where you actually want it to introduce it. The caller can do `.iter().rev().take(..)` for now.

I'm also not sure this is the *right* API. Perhaps this should take a height and give you all blocks >= that height and only the tip if that's empty. I don't know if you necessarily want `n` blocks because getting blocks that are 10,000 blocks deep already is a waste of time. I don't know the answer but I can't really decide here.

evanlinjin replied:
The problem with taking a height is you can't be sure if the method will return any blocks at all, or return a block that you can connect to.

The benefit of `max_count` is that it is the lower bound of the "height range size" returned (so you can use a number that is large enough to guarantee no reorgs). It also acts as the upper bound of elements to clone.

In any case, I agree that this is premature. Will remove.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -349,6 +349,11 @@ impl<A> TxGraph<A> {
     }
 
+    /// Iterates over all transaction anchors known by [`TxGraph`].
+    pub fn all_anchors(&self) -> impl ExactSizeIterator<Item = &(A, Txid)> + DoubleEndedIterator {
```
why not just return the `&BTreeSet<..>`. Much more useful in a `.anchors` method.

LLFourn reviewed (changes requested):
I left some comments. Other misc things I noticed that are not changed in this PR but could be:

1. `LocalChain::get_block(height) -> BlockId` allows for invalid representation. It should be `get_block_hash(height) -> BlockHash`.
2. It would be nice if `LocalChain` had a way to doing `.blocks` and just getting a reference to the inner BTreeMap without using the `AsRef` implementation.
3. There are all these `TODOs` as doc comments for Local chain. Can you make issues out of the TODOs and remove them from comments.

evanlinjin committed:
[chain_redesign] Various `LocalChain` improvements

* Introduce `LocalChain::inner` method to get the inner map of block
  height to hash.
* Replace `LocalChain::get_block` (which outputted `BlockId`, thus able
  to return invalid representation) with `get_blockhash` that just
returns a `BlockHash`.
* Remove `TODO` comments that should be github tickets.

danielabrozzoni commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -27,12 +28,14 @@ impl<A, I: Default> Default for IndexedTxGraph<A, I> {
 }
 
-impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I> {
+impl<A, I> IndexedTxGraph<A, I> {
```
I'm not sure how idiomatic this is, since we always end up using IndexedTxGraph with `A: Anchor, I: Indexer`, but I'm not opposed to this change if it's useful for us somehow

evanlinjin replied:
Even if it's not useful directly, it may allow building logic around/above it not require as many generic constraints. I would leave it as is unless if there are downsides to this?

danielabrozzoni commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -694,33 +694,43 @@ fn test_chain_spends() {
         .for_each(|(ht, tx)| {
-            let block_id = local_chain.get_block(*ht).expect("block expected");
-            let _ = graph.insert_anchor(tx.txid(), block_id);
+            // let block_id = local_chain.get_block(*ht).expect("block expected");
```
This can be removed?

danielabrozzoni commented on crates/chain/tests/test_indexed_tx_graph.rs:
```diff
@@ -206,86 +207,112 @@ fn test_list_owned_txouts() {
+                    local_chain
+                        .get_blockhash(height)
+                        .map(|hash| BlockId { height, hash })
+                        .unwrap(),
```
nit: just save this in a `tip` variable instead of repeating it over and over again (unless it's changing between calls, but it doesn't seem so?)

danielabrozzoni reviewed (approved):
ACK a56d289eef0ea241bfe29b4c1069966353e4ca57

A few comments but they don't have to be addressed

evanlinjin commented:
After a call with @LLFourn, the follow things will be added (either to this PR, or a separate PR):

1. The `PersistBackend<T, C>` trait currently has a generic `T` for the "in memory representation". This also constrains `C` to "accompany" `T`, thus requiring us to have something like `Tracker` in #965. The "onioning" of these structures is unfortunate. However, there is a way to do `PersistBackend` without `T` and just rely on `C`! Let's experiment with that to see if it cleans up the API and reduces "onioning".

2. `ChainOracle::get_chain_tip(&self) -> Result<Option<BlockId>, Self::Error>` will be added. This is needed as we need something to input as the "static block" for the `is_block_in_best_chain` call. Additionally, a structure that is capable of implementing `ChainOracle` functionality should also have the functionality to return the tip of the best chain.

3. `OwnedIndexer` is bad. Here is something better:

    ```rust
    pub OwnedTxOuts {
        SpkIndex;
        Iter: Iterator<Item = (Self::SpkIndex, OutPoint)>;
        fn owned_txouts(&self) -> Self::Iter;
    }
    ```
    This thing iterates over all outpoints that we own (alongside the spk index we give it). To get the owned UTXO set, we filter over this. Same thing for balance.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -71,18 +64,21 @@ impl LocalChain {
     }
 
+    /// Get a reference to the inner map of block height to hash.
+    pub fn inner(&self) -> &BTreeMap<u32, BlockHash> {
```
just call this `blocks` it's not really a wrapper so `inner` doesn't make much sense.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -71,18 +64,21 @@ impl LocalChain {
-            .map(|&hash| BlockId { height, hash })
+    /// Get a [`BlockHash`] at the given height.
+    pub fn get_blockhash(&self, height: u32) -> Option<BlockHash> {
+        self.blocks.get(&height).cloned()
```
oh right since we already expose the `BTreeMap` we don't even need this at all?

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -173,6 +169,31 @@ impl LocalChain {
     pub fn heights(&self) -> BTreeSet<u32> {
         self.blocks.keys().cloned().collect()
     }
+
```
same with the above we don't need it just expose the `BTreeMap`. Essentially LocalChain is just keeping a consistent BTreeMap so it only needs insert methods?

LLFourn reviewed (approved):
Left another idea. Approved. Happy to merge when @evanlinjin is happy.

evanlinjin committed:
[chain_redesign] Rename `LocalChain::inner()` to `blocks()`

Also, we can get rid of `LocalChain::get_blockhash`, since we can
already expose the internal map.

Additionally, tests and docs are improved.
